### DISCUSSION TITLE
Understanding if a maximum likelihood problem can be solved with CVXPY

### DISCUSSION THREAD

**Discussion by CarloNicolini** [cvxpy/cvxpy#1352] [Q&A] [answered]
I am trying  to solve a a maximum likelihood problem (maximum entropy graph ensembles), and I need to determine if, at least in this case with just 2 variables, CVXPY is the correct tool to estimate the parameters.
Currently the problem has the following structure (here only 2 variables for simplicity):

    x=cp.Variable(shape=(2,), pos=True, name='x')
    loglike = cp.log(x[0]) + cp.log(x[1]) - cp.log1p(x[0]*x[1])
    problem = cp.Problem(cp.Maximize(loglike),[x>=0, x<=1])

The first two terms of `loglike` are clearly concave (logarithms) while the second term `-cp.log1p(x[0]*x[1])` is QUASICONVEX and NONNEGATIVE, hence I am dealing with a `(CONCAVE,UNKNOWN)` + `(QUASICONVEX,NONNEGATIVE)` objective function, and the expression evaluates to `(UNKNOWN,UNKNOWN)`.
Can I exploit the powerful arsenal of CVXPY tools and optimizers to tackle this class of problems, or is this objective function out of target for CVXPY?

**Comment by SteveDiamond** [comment, marked as answer]
Hi Carlo, you could model the problem above as convex through the change of variables x = e^z. Then you get
```
z=cp.Variable(shape=(2,), name='log_x')
loglike = z[0] + z[1] - cp.log_sum_exp(cp.hstack([1, z[0]+z[1]]))
problem = cp.Problem(cp.Maximize(loglike),[z<=np.log(1)])
```
which is DCP.

**Reply by CarloNicolini** [reply]
Thanks Steve, this solution looks great and has a deep relation with the original problem. Indeed the variables `x` where obtained from the original ones with a change of variable `x=e^{-theta_i}` where `theta` are the Lagrangian multipliers of the maximum entropy problem. The `x` variables were supposed to simplify the expression of the log-likelihood, and they do it indeed, but at the expenses of making the problem no longer DCP. As a side note, I numerically verified that in the `x` variables the problem has negative definite hessian everywhere, so why should we perform this substitution? Do we need the log-log concave property?

**Reply by SteveDiamond** [reply]
The DCP system is a sufficient but not necessary condition for convexity. It is used in cvxpy because it is easy to verify and establishes important properties (besides convexity) used in the transformation into standard form.
